gridstatus/pjm.py

```diff
@@ -538,6 +603,447 @@ def _get_key(self):
+            params,
+            verbose=verbose,
+        )
+        update_docs = self._parse_xml_find_all(response.content, "update")
+        source_id = None
+        checkboxes = {}
+        for update_doc in update_docs:
+            checkbox_elems = update_doc.find_all("input", {"type": "checkbox"})
+            for checkbox_elem in checkbox_elems:
+                checked = checkbox_elem.get("checked", None) is not None
+                checkbox_name = checkbox_elem.get("name")
+                matches = re.search(r":([0-9]+):(j_.+)$", checkbox_name)
+                if matches is not None:
+                    to_check_idx = int(matches.group(1))
+                    if source_id is None:
+                        source_id = re.sub(r"_input$", "", matches.group(2))
+                checkboxes[to_check_idx] = checked
+
+        return (
+            checkboxes,
+            source_id,
+        )
+
+    def _dv_lmp_select_checkbox(
+        self,
+        dv_session,
+        form_source_id,
+        checkboxes,
+        to_check_idx,
+        verbose=False,
+    ):
+        # toggle checkbox
+        params = {
+            "javax.faces.partial.ajax": "true",
+            "javax.faces.source": f"chart1FrmLmpSelection:tblBusAggregates:{to_check_idx}:{form_source_id}",  # noqa: E501
+            "javax.faces.partial.execute": f"chart1FrmLmpSelection:tblBusAggregates:{to_check_idx}:{form_source_id}",  # noqa: E501
+            "javax.faces.partial.render": "globalMessages",
+            "javax.faces.behavior.event": "change",
+            "javax.faces.partial.event": "change",
+            "chart1FrmLmpSelection": "chart1FrmLmpSelection",
+            "chart1FrmLmpSelection:tblBusAggregates:name:filter": "",
+            "chart1FrmLmpSelection:tblBusAggregates:voltage:filter": "",
+            "chart1FrmLmpSelection:tblBusAggregates:station:filter": "",
+            "chart1FrmLmpSelection:tblBusAggregates_scrollState": "0,0",
+        }
+        checkboxes[to_check_idx] = True
+        for idx, is_checked in checkboxes.items():
+            if is_checked:
+                params[
+                    f"chart1FrmLmpSelection:tblBusAggregates"
+                    f":{idx}:{form_source_id}_input"  # noqa: E501
+                ] = "on"
+        response = self._dv_lmp_fetch(
+            dv_session,
+            params,
+            verbose=verbose,
+        )
+        if verbose:
+            print(f"response (truncated) = {response.content[0:256]}", file=sys.stderr)
+
+    def _get_lmp_via_dv(
+        self,
+        date,
+        market: str,
+        end=None,
+        locations="hubs",
+        location_type=None,
+        verbose=False,
+    ):
+        """Get latest LMP data from Data Viewer, which includes RT & DA"""
+        with requests.session() as session:
+            initial_fetch = session.get(DATAVIEWER_LMP_URL)
+
+            dv_session = self._new_dv_session(session, verbose=verbose)
+            if dv_session is None:
```

Comment:
i would raise this error inside `_new_dv_session`. I dont see advantage to doing it here. by moving it, it keeps this function more readable. same with `chart_ids is None:` and `_dv_lmp_extract_chart_ids` below